examples/llama/main.zig

```diff
@@ -1,156 +1,228 @@
+        // current token index needs to go into a zml.Buffer
+        const token_index_buffer = &[_]i32{@intCast(prompt_tok.len + i)};
+        const token_index = try zml.Buffer.fromSlice(platform, .{}, token_index_buffer);
+        defer token_index.deinit();
+
+        // call to generate the next token
+        current_token, kv_cache, rng = mod_generate.call(.{ current_token, token_index, kv_cache, rng });
+
+        tracer.frameEnd(frame_id, try std.fmt.bufPrintZ(tracer_buffer, "Generated token {}/{}", .{ i + 1, output_tokens_len }));
+
+        // extract the generated token from the buffer, async
+        var fut_tok_buf = try current_token.toHost(std.mem.sliceAsBytes(&generated_token_buffer));
+        _ = try fut_tok_buf.awaitt();
+        const generated_token = generated_token_buffer[0];
+
+        // de-tokenize generated token into a string
+        const chunk = try tokenizer_decoder.next(@intCast(generated_token)) orelse unreachable;
+        num_tokens_generated = i;
+
+        // check for eos
+        switch (config.eos_token_id.value) {
+            .int => |eos| if (generated_token == @as(u32, @intCast(eos))) break :generation,
+            .ints => |eos_list| {
+                for (eos_list) |eos| {
+                    if (generated_token == @as(u32, @intCast(eos))) break :generation;
+                }
+            },
         }
+
+        // collect and print generated sequence
+        try output.appendSlice(chunk);
+        std.debug.print("{s}", .{chunk});
     }
-    var total_token_count: usize = max_seq_len;
-    const n = output.items.len;
-    if (eos_index) |end_idx| {
-        // count = eos index + 1
-        total_token_count = end_idx + 1;
-    }
-    const generated_token_count = total_token_count - prompt_tok.len;
-    try tokenizer.decodeWithOpts(&output, @ptrCast(token_buffer[decode_progress..total_token_count]), .{});
-    std.debug.print("{s}\n", .{output.items[n..]});
     const end = std.time.microTimestamp();
-
     const duration = stdx.math.divFloat(f64, end - start, std.time.us_per_s);
-    const speed = @as(f64, @floatFromInt(generated_token_count)) / duration;
-    log.info("✅ Generated {d} tokens in {:.3}s: {d:.3}tok/s", .{ generated_token_count, duration, speed });
-
-    _ = try tokens.toHost(std.mem.sliceAsBytes(token_buffer));
-    output.clearRetainingCapacity();
-
-    try tokenizer.decodeWithOpts(&output, @ptrCast(token_buffer[0..total_token_count]), .{});
+    const speed = @as(f64, @floatFromInt(num_tokens_generated)) / duration;
+    std.debug.print("\n", .{});
+    log.info("✅ Generated {d} tokens in {:.3}s: {d:.3}tok/s", .{ num_tokens_generated, duration, speed });
     return output.toOwnedSlice();
 }
 
+const params = clap.parseParamsComptime(
+    \\--help                    print this help
+    \\--prompt <STRING>         the prompt
+    \\--model-config <PATH>     config.json path
+    \\--model-name <STRING>     model name
+    \\--model-weights <PATH>    model weights path
```

Comment:
the model prefix make it harder to type the cli, and the code harder to read because of the `args.@"model-tokenizer"`. not sure it's worth it.

Reply:
it isn't here. we don't have many groups of params. so I'll drop it.